src/Orleans.Serialization/Codecs/IntegerCodec.cs

```diff
@@ -745,4 +749,210 @@ public static ulong ReadValue<TInput>(ref Reader<TInput> reader, Field field)
+        /// <param name="value">The value.</param>
+        public static void WriteField<TBufferWriter>(ref Writer<TBufferWriter> writer, uint fieldIdDelta, Type expectedType, UInt128 value) where TBufferWriter : IBufferWriter<byte>
+        {
+            if (value <= (UInt128)ulong.MaxValue && value >= (UInt128)ulong.MinValue)
+            {
+                UInt64Codec.WriteField(ref writer, fieldIdDelta, expectedType, (ulong)value, CodecFieldType);
+            }
+            else
+            {
+                ReferenceCodec.MarkValueField(writer.Session);
+                UnsignedBinaryIntegerCodec<UInt128>.WriteField(ref writer, fieldIdDelta, expectedType, in value);
+            }
+        }
+
+        /// <inheritdoc/>
+        UInt128 IFieldCodec<UInt128>.ReadValue<TInput>(ref Reader<TInput> reader, Field field) => ReadValue(ref reader, field);
+
+        /// <summary>
+        /// Reads a value.
+        /// </summary>
+        /// <typeparam name="TInput">The reader input type.</typeparam>
+        /// <param name="reader">The reader.</param>
+        /// <param name="field">The field.</param>
+        /// <returns>The value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static UInt128 ReadValue<TInput>(ref Reader<TInput> reader, Field field)
+        {
+            if (field.WireType != WireType.LengthPrefixed)
+            {
+                return UInt64Codec.ReadValue(ref reader, field);
+            }
+
+            ReferenceCodec.MarkValueField(reader.Session);
+            UInt128 result;
+            UnsignedBinaryIntegerCodec<UInt128>.ReadLengthPrefixed(ref reader, out result);
+            return result;
+        }
+
+        /// <summary>
+        /// Reads a value without protocol framing.
+        /// </summary>
+        /// <typeparam name="TInput">The reader input type.</typeparam>
+        /// <param name="reader">The reader.</param>
+        /// <returns>The value.</returns>
+        internal static UInt128 ReadRaw<TInput>(ref Reader<TInput> reader)
+        {
+            UInt128 result;
+            UnsignedBinaryIntegerCodec<UInt128>.ReadLengthPrefixed(ref reader, out result);
+            return result;
+        }
+    }
+
+    internal static class SignedBinaryIntegerCodec<T> where T : struct, IBinaryInteger<T>, ISignedNumber<T>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteField<TBufferWriter>(scoped ref Writer<TBufferWriter> writer, uint fieldIdDelta, Type expectedType, scoped in T value) where TBufferWriter : IBufferWriter<byte>
+        {
+            var byteCount = value.GetByteCount();
+            writer.WriteFieldHeader(fieldIdDelta, expectedType, typeof(T), WireType.LengthPrefixed);
+            writer.WriteVarUInt32((uint)byteCount);
+            Span<byte> bytes = byteCount < 64 ? stackalloc byte[64].Slice(0, byteCount) : new byte[byteCount];
+            value.WriteLittleEndian(bytes);
```

Comment:
> This code unfortunately seems to allocate.

Stripped down repro:
```c#
using System.Buffers;
using System.Numerics;
using System.Runtime.CompilerServices;

ArrayBufferWriter<byte> writer = new();
for (int i = 0; i < 1_000_000; ++i)
{
    SignedBinaryIntegerCodec<Int128>.WriteField(writer, Int128.MinValue);
    writer.Clear();
}

Console.WriteLine(GC.CollectionCount(0));

internal static class SignedBinaryIntegerCodec<T> where T : struct, IBinaryInteger<T>
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void WriteField<TBufferWriter>(TBufferWriter writer, scoped in T value)
        where TBufferWriter : IBufferWriter<byte>
    {
        int byteCount = value.GetByteCount();
        Span<byte> span = writer.GetSpan(byteCount);
        value.WriteLittleEndian(span);
        writer.Advance(byteCount);
    }
}
```

Solution:
```suggestion
            value.TryWriteLittleEndian(bytes, out var written);
            Debug.Assert(written == byteCount);
```

Explanation:
Both `WriteLittleEndian` and `TryWriteLittleEndian` are defined in `IBinaryInteger<T>`.
All the numeric types [implement TryWriteLittleEndian explicitely](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryInteger.cs,1af7c8501ad6bd32,references), while none of the implementes `WriteLittleEndian` which is implemented as default interface method.
Thus the call `value.WriteLittleEndian`, when `value` is a value-type, needs to be boxed to the interface and this is the allocations you see.

It's a bit unfortunate that the implementation in runtime-repo is as it is. Maybe we should open an issue over there.

Reply:
Opened https://github.com/dotnet/runtime/issues/77969 for this.

Reply:
Thanks for investigating, @gfoidl